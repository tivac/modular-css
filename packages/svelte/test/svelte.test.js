"use strict";

const fs = require("fs");
const path = require("path");

const svelte = require("svelte");
const dedent = require("dedent");

const Processor = require("@modular-css/processor");
const namer = require("@modular-css/test-utils/namer.js");
const logs  = require("@modular-css/test-utils/logs.js");

const plugin = require("../svelte.js");

describe("/svelte.js", () => {
    afterEach(() => require("shelljs").rm("-rf", "./packages/svelte/test/output/*"));

    it("should extract CSS from a <style> tag", async () => {
        const filename = require.resolve("./specimens/style.html");
        const { processor, preprocess } = plugin({
            namer,
        });

        const processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();

        const output = await processor.output();

        expect(output.css).toMatchSnapshot();
    });
    
    it("should ignore <links> that reference a URL", async () => {
        const filename = require.resolve("./specimens/url.html");
        const { preprocess, processor } = plugin({
            namer,
        });

        const processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();

        const output = await processor.output();

        expect(output.css).toMatchSnapshot();
    });

    it("should use an already-created processor", async () => {
        const processor = new Processor({ namer });

        await processor.string(
            "./fake.css",
            ".fake { color: #F00; }"
        );

        const filename = require.resolve("./specimens/url.html");
        const { preprocess } = plugin({
            processor,
        });

        const processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();

        const output = await processor.output();

        expect(output.css).toMatchSnapshot();
    });

    it.each`
        specimen                    | title
        ${"external.html"}          | ${"no script"}
        ${"external-script.html"}   | ${"existing script"}
        ${"external-single.html"}   | ${"single quotes"}
        ${"external-unquoted.html"} | ${"unquoted"}
        ${"external-values.html"}   | ${"values"}
    `("should extract CSS from a <link> tag ($title)", async ({ specimen }) => {
        const filename = require.resolve(`./specimens/${specimen}`);
        const { processor, preprocess } = plugin({
            namer,
        });

        const processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();

        const output = await processor.output();

        expect(output.css).toMatchSnapshot();
    });

    it("should ignore files without <style> blocks", async () => {
        const { processor, preprocess } = plugin();

        const processed = await svelte.preprocess(
            dedent(`
                <h1>Hello</h1>
                <script>console.log("output")</script>
            `),
            preprocess
        );

        expect(processed.toString()).toMatchSnapshot();

        const output = await processor.output();

        expect(output.css).toMatchSnapshot();
    });

    it.each`
        title                                     | specimen
        ${"invalid reference <script> - <style>"} | ${"invalid-inline-script.html"}
        ${"invalid reference template - <style>"} | ${"invalid-inline-template.html"}
        ${"invalid reference <script> - <link>"}  | ${"invalid-external-script.html"}
        ${"invalid reference template - <link>"}  | ${"invalid-external-template.html"}
        ${"empty css file - <style>"}             | ${"invalid-inline-empty.html"}
        ${"empty css file - <link>"}              | ${"invalid-external-empty.html"}
    `("should handle errors: $title", async ({ specimen }) => {
        const spy = jest.spyOn(global.console, "warn");

        spy.mockImplementation(() => { /* NO-OP */ });

        const filename = require.resolve(`./specimens/${specimen}`);

        // Set up strict plugin
        const { preprocess : strict } = plugin({
            namer,
            strict : true,
        });

        await expect(
            svelte.preprocess(
                fs.readFileSync(filename, "utf8"),
                Object.assign({}, strict, { filename })
            )
        ).rejects.toThrowErrorMatchingSnapshot();

        // Now the loose plugin
        const { preprocess : loose } = plugin({
            namer,
        });

        const processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, loose, { filename })
        );

        expect(spy).toHaveBeenCalled();
        expect(spy.mock.calls).toMatchSnapshot();

        expect(processed.toString()).toMatchSnapshot();
    });

    it("should throw on both <style> and <link> in one file", async () => {
        const { preprocess } = plugin({
            css : "./packages/svelte/test/output/svelte.css",
            namer,
        });

        const filename = require.resolve("./specimens/both.html");

        await expect(
            svelte.preprocess(
                fs.readFileSync(filename, "utf8"),
                Object.assign({}, preprocess, { filename })
            )
        ).rejects.toThrowErrorMatchingSnapshot();
    });

    it.each`
        title        | specimen
        ${"<style>"} | ${"style.html"}
        ${"<link>"}  | ${"external.html"}
    `("should support verbose output: $title", async ({ specimen }) => {
        const { logSnapshot } = logs();

        const filename = require.resolve(`./specimens/${specimen}`);

        const { processor, preprocess } = plugin({
            namer,
            verbose : true,
        });

        await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        await processor.output();

        logSnapshot();
    });

    it("should warn when multiple <link> elements are in the html", async () => {
        const spy = jest.spyOn(global.console, "warn");

        spy.mockImplementation(() => { /* NO-OP */ });

        const filename = require.resolve(`./specimens/multiple-link.html`);

        const { processor, preprocess } = plugin({
            namer,
        });

        const processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();

        const output = await processor.output();
        
        expect(output.css).toMatchSnapshot();
        
        expect(spy).toHaveBeenCalled();
        expect(spy.mock.calls).toMatchSnapshot();
    });

    it("should no-op if all <link>s reference a URL", async () => {
        const filename = require.resolve("./specimens/multiple-url.html");
        
        const { preprocess } = plugin({
            namer,
        });

        const processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();
    });
    

    it("should invalidate files before reprocessing (<link>)", async () => {
        // V1 of files
        fs.writeFileSync(path.resolve(__dirname, "./output/source.html"), dedent(`
            <link rel="stylesheet" href="./source.css" />
            <div class="{css.source}">Source</div>
        `));

        fs.writeFileSync(path.resolve(__dirname, "./output/source.css"), dedent(`
            .source {
                color: red;
            }
        `));

        const filename = require.resolve(`./output/source.html`);
        const { processor, preprocess } = plugin({
            namer,
        });

        let processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();

        let output = await processor.output();

        expect(output.css).toMatchSnapshot();

        // V2 of CSS
        fs.writeFileSync(path.resolve(__dirname, "./output/source.css"), dedent(`
            .source {
                color: blue;
            }
        `));

        processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();

        output = await processor.output();

        expect(output.css).toMatchSnapshot();
    });

    it("should invalidate files before reprocessing (<style>)", async () => {
        // V1 of files
        fs.writeFileSync(path.resolve(__dirname, "./output/source.html"), dedent(`
            <style>.source { color: red; }</style>
            <div class="{css.source}">Source</div>
        `));

        const filename = require.resolve(`./output/source.html`);
        const { processor, preprocess } = plugin({
            namer,
        });

        let processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();

        let output = await processor.output();

        expect(output.css).toMatchSnapshot();

        // V2 of CSS
        fs.writeFileSync(path.resolve(__dirname, "./output/source.html"), dedent(`
            <style>.source { color: blue; }</style>
            <div class="{css.source}">Source</div>
        `));

        processed = await svelte.preprocess(
            fs.readFileSync(filename, "utf8"),
            Object.assign({}, preprocess, { filename })
        );

        expect(processed.toString()).toMatchSnapshot();

        output = await processor.output();

        expect(output.css).toMatchSnapshot();
    });

    it("should wait for files to finish", async () => {
        const { preprocess } = plugin({
            namer,
        });

        const results = await Promise.all(
            [
                require.resolve("./specimens/overlapping/entry1.html"),
                require.resolve("./specimens/overlapping/entry2.html"),
            ]
            .map((filename) => svelte.preprocess(
                fs.readFileSync(filename, "utf8"),
                Object.assign({}, preprocess, { filename })
            ))
        );

        expect(results.map((result) => result.toString())).toMatchSnapshot();
    });
});
