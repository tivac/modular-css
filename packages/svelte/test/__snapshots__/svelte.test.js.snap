// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/svelte.js should extract CSS from a <link> tag ("existing script") 1`] = `
"
<div class=\\"flex wrapper\\">
    <h1 class=\\"flex fooga hd\\">Head</h1>
    <div class=\\"fooga wooga bd\\">
        <p class=\\"{bool ? \\"text\\" : \\"active\\" }\\">Text</p>
    </div>
</div>

<script>
import css from \\"./external.css\\";


export default {
    data : () => ({
        bool : true
    }),

    oncreate() {
        const bd = {
            flex : \\"flex\\"
        };

        const style = css[bd.flex];
    }
};
</script>
"
`;

exports[`/svelte.js should extract CSS from a <link> tag ("existing script") 2`] = `
"/* packages/svelte/test/specimens/simple.css */
.fooga {
    color: red;
}
/* packages/svelte/test/specimens/dependencies.css */
.wooga {

    background: blue;
}
/* packages/svelte/test/specimens/external.css */
.flex {
    display: flex;
}
.text {
    color: #000;
}
.active {
    color: #F00;
}
"
`;

exports[`/svelte.js should extract CSS from a <link> tag ("no script") 1`] = `
"
<div class=\\"flex wrapper\\">
    <h1 class=\\"flex fooga hd\\">Head</h1>
    <div class=\\"fooga wooga bd\\">
        <p class=\\"{bool ? \\"text\\" : \\"active\\" }\\">Text</p>
    </div>
</div>
<script>
    import css from \\"./external.css\\";
</script>"
`;

exports[`/svelte.js should extract CSS from a <link> tag ("no script") 2`] = `
"/* packages/svelte/test/specimens/simple.css */
.fooga {
    color: red;
}
/* packages/svelte/test/specimens/dependencies.css */
.wooga {

    background: blue;
}
/* packages/svelte/test/specimens/external.css */
.flex {
    display: flex;
}
.text {
    color: #000;
}
.active {
    color: #F00;
}
"
`;

exports[`/svelte.js should extract CSS from a <link> tag ("single quotes") 1`] = `
"
<div class=\\"flex wrapper\\">
    <h1 class=\\"flex fooga hd\\">Head</h1>
    <div class=\\"fooga wooga bd\\">
        <p class=\\"{bool ? \\"text\\" : \\"active\\" }\\">Text</p>
    </div>
</div>
<script>
    import css from \\"./external.css\\";
</script>"
`;

exports[`/svelte.js should extract CSS from a <link> tag ("single quotes") 2`] = `
"/* packages/svelte/test/specimens/simple.css */
.fooga {
    color: red;
}
/* packages/svelte/test/specimens/dependencies.css */
.wooga {

    background: blue;
}
/* packages/svelte/test/specimens/external.css */
.flex {
    display: flex;
}
.text {
    color: #000;
}
.active {
    color: #F00;
}
"
`;

exports[`/svelte.js should extract CSS from a <link> tag ("unquoted") 1`] = `
"
<div class=\\"flex wrapper\\">
    <h1 class=\\"flex fooga hd\\">Head</h1>
    <div class=\\"fooga wooga bd\\">
        <p class=\\"{bool ? \\"text\\" : \\"active\\" }\\">Text</p>
    </div>
</div>
<script>
    import css from \\"./external.css\\";
</script>"
`;

exports[`/svelte.js should extract CSS from a <link> tag ("unquoted") 2`] = `
"/* packages/svelte/test/specimens/simple.css */
.fooga {
    color: red;
}
/* packages/svelte/test/specimens/dependencies.css */
.wooga {

    background: blue;
}
/* packages/svelte/test/specimens/external.css */
.flex {
    display: flex;
}
.text {
    color: #000;
}
.active {
    color: #F00;
}
"
`;

exports[`/svelte.js should extract CSS from a <link> tag ("values") 1`] = `
"
<div
    data-simple=\\"color: #BEEFED\\"
    data-expression=\\"{\\"color: \\" + \\"#BEEFED\\"}\\"
    data-template=\\"{\`color: \${\\"#BEEFED\\"}\`}\\"
>
</div>

<script>
import css from \\"./external.css\\";


    export default {
        data : () => ({
            template : \`color: \${\\"#BEEFED\\"}\`,
            expression : \\"color: \\" + \\"#BEEFED\\",
        })
    };
</script>
"
`;

exports[`/svelte.js should extract CSS from a <link> tag ("values") 2`] = `
"/* packages/svelte/test/specimens/simple.css */
.fooga {
    color: red;
}
/* packages/svelte/test/specimens/dependencies.css */
.wooga {

    background: blue;
}
/* packages/svelte/test/specimens/external.css */
.flex {
    display: flex;
}
.text {
    color: #000;
}
.active {
    color: #F00;
}
"
`;

exports[`/svelte.js should extract CSS from a <style> tag 1`] = `
"<div class=\\"flex wrapper\\">
    <h1 class=\\"flex fooga hd\\">Head</h1>
    <div class=\\"fooga wooga bd\\">
        <p class=\\"{bool ? \\"text\\" : \\"active\\" }\\">Text</p>
    </div>
</div>

<style>/* replaced by modular-css */</style>

<script>
export default {
    data : () => ({
        bool : true
    })
};
</script>
"
`;

exports[`/svelte.js should extract CSS from a <style> tag 2`] = `
"/* packages/svelte/test/specimens/simple.css */
.fooga {
    color: red;
}
/* packages/svelte/test/specimens/dependencies.css */
.wooga {

    background: blue;
}
/* packages/svelte/test/specimens/style.html */
.flex {
        display: flex;
    }
.text {
        color: #000;
    }
.active {
        color: #F00;
    }
"
`;

exports[`/svelte.js should handle errors: "empty css file - <link>" 1`] = `"@modular-css/svelte: Unable to find .nope, .alsonope in \\"./empty.css\\""`;

exports[`/svelte.js should handle errors: "empty css file - <link>" 2`] = `
Array [
  Array [
    "@modular-css/svelte: Unable to find .nope in \\"./empty.css\\"",
  ],
  Array [
    "@modular-css/svelte: Unable to find .alsonope in \\"./empty.css\\"",
  ],
]
`;

exports[`/svelte.js should handle errors: "empty css file - <link>" 3`] = `
"
<div class=\\"{css.nope}\\">NOPE</div>
<div class=\\"{css.alsonope}\\">STILL NOPE</div>
<script>
    import css from \\"./empty.css\\";
</script>"
`;

exports[`/svelte.js should handle errors: "empty css file - <style>" 1`] = `"@modular-css/svelte: Unable to find .nope, .alsonope in \\"<style>\\""`;

exports[`/svelte.js should handle errors: "empty css file - <style>" 2`] = `
Array [
  Array [
    "@modular-css/svelte: Unable to find .nope in \\"<style>\\"",
  ],
  Array [
    "@modular-css/svelte: Unable to find .alsonope in \\"<style>\\"",
  ],
]
`;

exports[`/svelte.js should handle errors: "empty css file - <style>" 3`] = `
"<div class=\\"{css.nope}\\">NOPE</div>
<div class=\\"{css.alsonope}\\">STILL NOPE</div>

<style>/* replaced by modular-css */</style>
"
`;

exports[`/svelte.js should handle errors: "invalid reference <script> - <link>" 1`] = `"@modular-css/svelte: Unable to find .nuhuh in \\"./invalid.css\\""`;

exports[`/svelte.js should handle errors: "invalid reference <script> - <link>" 2`] = `
Array [
  Array [
    "@modular-css/svelte: Unable to find .nuhuh in \\"./invalid.css\\"",
  ],
]
`;

exports[`/svelte.js should handle errors: "invalid reference <script> - <link>" 3`] = `
"
<h2 class=\\"yup\\">Yup</h2>

<script>
import css from \\"./invalid.css\\";


    console.log(css.nuhuh);
</script>
"
`;

exports[`/svelte.js should handle errors: "invalid reference <script> - <style>" 1`] = `"@modular-css/svelte: Unable to find .nuhuh in \\"<style>\\""`;

exports[`/svelte.js should handle errors: "invalid reference <script> - <style>" 2`] = `
Array [
  Array [
    "@modular-css/svelte: Unable to find .nuhuh in \\"<style>\\"",
  ],
]
`;

exports[`/svelte.js should handle errors: "invalid reference <script> - <style>" 3`] = `
"<h2 class=\\"yup\\">Yup</h2>

<style>/* replaced by modular-css */</style>

<script>
    console.log(css.nuhuh);
</script>
"
`;

exports[`/svelte.js should handle errors: "invalid reference template - <link>" 1`] = `"@modular-css/svelte: Unable to find .nope in \\"./invalid.css\\""`;

exports[`/svelte.js should handle errors: "invalid reference template - <link>" 2`] = `
Array [
  Array [
    "@modular-css/svelte: Unable to find .nope in \\"./invalid.css\\"",
  ],
]
`;

exports[`/svelte.js should handle errors: "invalid reference template - <link>" 3`] = `
"
<h1 class=\\"{css.nope}\\">Nope</h1>
<h2 class=\\"yup\\">Yup</h2>
<script>
    import css from \\"./invalid.css\\";
</script>"
`;

exports[`/svelte.js should handle errors: "invalid reference template - <style>" 1`] = `"@modular-css/svelte: Unable to find .nope in \\"<style>\\""`;

exports[`/svelte.js should handle errors: "invalid reference template - <style>" 2`] = `
Array [
  Array [
    "@modular-css/svelte: Unable to find .nope in \\"<style>\\"",
  ],
]
`;

exports[`/svelte.js should handle errors: "invalid reference template - <style>" 3`] = `
"<h1 class=\\"{css.nope}\\">Nope</h1>
<h2 class=\\"yup\\">Yup</h2>

<style>/* replaced by modular-css */</style>
"
`;

exports[`/svelte.js should ignore <links> that reference a URL 1`] = `
"<link rel=\\"stylesheet\\" href=\\"http://example.com/styles.css\\" />

<div class=\\"fooga\\">fooga</div>
<script>
    import css from \\"./simple.css\\";
</script>"
`;

exports[`/svelte.js should ignore <links> that reference a URL 2`] = `
"/* packages/svelte/test/specimens/simple.css */
.fooga {
    color: red;
}
"
`;

exports[`/svelte.js should ignore files without <style> blocks 1`] = `
"<h1>Hello</h1>
<script>console.log(\\"output\\")</script>"
`;

exports[`/svelte.js should ignore files without <style> blocks 2`] = `""`;

exports[`/svelte.js should invalidate files before reprocessing (<link>) 1`] = `
"<div class=\\"source\\">Source</div><script>
    import css from \\"./source.css\\";
</script>"
`;

exports[`/svelte.js should invalidate files before reprocessing (<link>) 2`] = `
"/* packages/svelte/test/output/source.css */
.source {
    color: red;
}"
`;

exports[`/svelte.js should invalidate files before reprocessing (<link>) 3`] = `
"<div class=\\"source\\">Source</div><script>
    import css from \\"./source.css\\";
</script>"
`;

exports[`/svelte.js should invalidate files before reprocessing (<link>) 4`] = `
"/* packages/svelte/test/output/source.css */
.source {
    color: blue;
}"
`;

exports[`/svelte.js should invalidate files before reprocessing (<style>) 1`] = `
"<style>/* replaced by modular-css */</style>
<div class=\\"source\\">Source</div>"
`;

exports[`/svelte.js should invalidate files before reprocessing (<style>) 2`] = `
"/* packages/svelte/test/output/source.html */
.source { color: red; }"
`;

exports[`/svelte.js should invalidate files before reprocessing (<style>) 3`] = `
"<style>/* replaced by modular-css */</style>
<div class=\\"source\\">Source</div>"
`;

exports[`/svelte.js should invalidate files before reprocessing (<style>) 4`] = `
"/* packages/svelte/test/output/source.html */
.source { color: blue; }"
`;

exports[`/svelte.js should no-op if all <link>s reference a URL 1`] = `
"<link rel=\\"stylesheet\\" href=\\"http://example.com/styles.css\\" />
<link rel=\\"stylesheet\\" href=\\"http://example.com/styles2.css\\" />

<div class=\\"fooga\\">fooga</div>
"
`;

exports[`/svelte.js should support verbose output: "<link>" 1`] = `
Array [
  Array [
    "[svelte]",
    "Processing",
    "packages/svelte/test/specimens/external.html",
  ],
  Array [
    "[svelte]",
    "extract <link>",
    "packages/svelte/test/specimens/external.css",
  ],
  Array [
    "[processor]",
    "file()",
    "packages/svelte/test/specimens/external.css",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/svelte/test/specimens/external.css",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/svelte/test/specimens/external.css",
  ],
  Array [
    "[processor]",
    "file()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "file()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/svelte/test/specimens/external.css",
  ],
  Array [
    "[svelte]",
    "updating source {css.<key>} references from",
    "./external.css",
  ],
  Array [
    "[svelte]",
    "[\\"color\\",\\"flex\\",\\"wrapper\\",\\"hd\\",\\"bd\\",\\"text\\",\\"active\\"]",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/svelte/test/specimens/external.css",
  ],
]
`;

exports[`/svelte.js should support verbose output: "<style>" 1`] = `
Array [
  Array [
    "[svelte]",
    "Processing",
    "packages/svelte/test/specimens/style.html",
  ],
  Array [
    "[svelte]",
    "extract <style>",
  ],
  Array [
    "[processor]",
    "string()",
    "packages/svelte/test/specimens/style.html",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/svelte/test/specimens/style.html",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/svelte/test/specimens/style.html",
  ],
  Array [
    "[processor]",
    "file()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "file()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_add()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_before()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_process()",
    "packages/svelte/test/specimens/style.html",
  ],
  Array [
    "[svelte]",
    "updating source {css.<key>} references from",
    "<style>",
  ],
  Array [
    "[svelte]",
    "[\\"flex\\",\\"wrapper\\",\\"hd\\",\\"bd\\",\\"text\\",\\"active\\"]",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/svelte/test/specimens/simple.css",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/svelte/test/specimens/dependencies.css",
  ],
  Array [
    "[processor]",
    "_after()",
    "packages/svelte/test/specimens/style.html",
  ],
]
`;

exports[`/svelte.js should throw on both <style> and <link> in one file 1`] = `"@modular-css/svelte: use <style> OR <link>, but not both"`;

exports[`/svelte.js should use an already-created processor 1`] = `
"<link rel=\\"stylesheet\\" href=\\"http://example.com/styles.css\\" />

<div class=\\"fooga\\">fooga</div>
<script>
    import css from \\"./simple.css\\";
</script>"
`;

exports[`/svelte.js should use an already-created processor 2`] = `
"/* fake.css */
.fake { color: #F00; }
/* packages/svelte/test/specimens/simple.css */
.fooga {
    color: red;
}"
`;

exports[`/svelte.js should warn when multiple <link> elements are in the html 1`] = `
"<link rel=\\"stylesheet\\" href=\\"./simple2.css\\" />

<div class=\\"fooga\\">fooga</div>
<div class=\\"{css.booga}\\">booga</div>
<script>
    import css from \\"./simple.css\\";
</script>"
`;

exports[`/svelte.js should warn when multiple <link> elements are in the html 2`] = `
"/* packages/svelte/test/specimens/simple.css */
.fooga {
    color: red;
}
"
`;

exports[`/svelte.js should warn when multiple <link> elements are in the html 3`] = `
Array [
  Array [
    "@modular-css/svelte will only use the first local <link> tag",
  ],
  Array [
    "@modular-css/svelte: Unable to find .booga in \\"./simple.css\\"",
  ],
]
`;
